shaders/oit_linked_lists/combine.frag

```diff
@@ -0,0 +1,106 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#define LINKED_LIST_END_SENTINEL	0xFFFFFFFFU
+#define SORTED_FRAGMENT_MAX_COUNT	16U
+
+////////////////////////////////////////
+
+layout(set = 0, binding = 0) uniform SceneConstants
+{
+	mat4 projection;
+	mat4 view;
+	uvec2 unused;
+	uint sortFragments;
+	uint fragmentMaxCount;
+} sceneConstants;
+
+layout(set = 0, binding = 2, r32ui) uniform uimage2D linkedListHeadTex;
+
+layout(set = 0, binding = 3) buffer FragmentBuffer {
+	uvec3 data[];
+} fragmentBuffer;
+
+layout(set = 0, binding = 4) buffer FragmentCounter {
+	uint value;
+} fragmentCounter;
+
+////////////////////////////////////////
+
+layout (location = 0) out vec4 outFragColor;
+
+////////////////////////////////////////
+
+void main()
+{
+    // Reset the atomic counter for the next frame
+	// Note that we don't care about atomicity here, as all threads will write the same value
+    fragmentCounter.value = 0;
+
+    // Get the first fragment index in the linked list
+	uint fragmentIndex = imageLoad(linkedListHeadTex, ivec2(gl_FragCoord.xy)).r;
+	// Reset the list head for the next frame
+	imageStore(linkedListHeadTex, ivec2(gl_FragCoord.xy), uvec4(LINKED_LIST_END_SENTINEL, 0, 0, 0));
+
+    // Copy the fragments into local memory for sorting
+    uvec2 sortedFragments[SORTED_FRAGMENT_MAX_COUNT];
+    uint fragmentCount = 0U;
+    while(fragmentIndex != LINKED_LIST_END_SENTINEL && fragmentCount < SORTED_FRAGMENT_MAX_COUNT)
+    {
+        const uvec3 fragment = fragmentBuffer.data[fragmentIndex];
+        fragmentIndex = fragment.x;
+        sortedFragments[fragmentCount] = fragment.yz;
+        ++fragmentCount;
+    }
+
+	// Early return if there are no fragments
+    if(fragmentCount == 0)
+    {
+		outFragColor = vec4(0.0f);
+        return;
+    }
+
+    // Sort the fragments by depth (back to front)
+	if(sceneConstants.sortFragments == 1U)
+    {
+        for(int i = 0; i < fragmentCount - 1; ++i)
+        {
+            for(int j = i + 1; j < fragmentCount; ++j)
+            {
+                if(sortedFragments[j].y < sortedFragments[i].y)
+                {
+                    const uvec2 tmp = sortedFragments[i];
+                    sortedFragments[i] = sortedFragments[j];
+                    sortedFragments[j] = tmp;
+                }
+            }
+        }
+    }
+
+    // Merge the fragments to get the final color
+    vec4 color = vec4(0.0f, 0.0f, 0.0f, 1.0f);
+    for(int i = 0; i < fragmentCount; ++i)
+    {
+		const vec4 fragmentColor = unpackUnorm4x8(sortedFragments[i].x);
+		color.rgb = mix(color.rgb, fragmentColor.rgb, fragmentColor.a);
```

Comment:
I've personally always found standard alpha blending looks very unrealistic. Since you're actually performing all the blend calculations here you could actually make use of properly filtered colour blending (i.e. the way that glass or gel films actually behave).

Here's something I hacked in to do that (using another checkbox to control filtered vs standard).
Filtering the colours is just essentially multiplying src colour and dst colour. The opacity (alpha) then just controls how much you mix between fully transparent (just src*dst) and fully opaque (just src).
You have the alpha from 0.2 to 1.0 in the sample, but this kind of filtering actually looks nicer biased towards lower alpha values.

Try it, see if you like it. I can attach a video of it running also.

```
// Merge the fragments to get the final color
vec4 color = vec4(0.3f, 0.3f, 0.3f, 1.0f);
for(int i = 0; i < fragmentCount; ++i)
{
    const vec4 fragmentColor = unpackUnorm4x8(sortedFragments[i].x);
     vec3 first = color.rgb;
     if (sceneConstants.filtered == 1U)
         first *= fragmentColor.rgb;
    color.rgb = mix(first, fragmentColor.rgb, fragmentColor.a);
    color.a *= 1.0f - fragmentColor.a;
    }
if (sceneConstants.filtered == 1U)
    color.a = 1.0f;
else
    color.a = 1.0f - color.a;
outFragColor = color;
```

Reply:
Oh, excellent point!
I completely forgot to point out that the algorithm gives more freedom over the blending equation.

I added support for your equation and a corresponding checkbox.
I also updated the README to point out that the blending equation can be customized.
Could you please have a look to make sure the rendering is correct with the new option?

Reply:
> > Could you please have a look to make sure the rendering is correct with the new option?
> 
> It doesn't look quite right as it stands. I believe you'll need to disable the native blending when using the filtered option. Anything that just blocks the background looks wrong otherwise. You can do this by just setting color.a to 1.0 in the filtered case.
> 
> ```
> if (sceneConstants.filtered_blending == 1U)
>     color.a = 1.0f;
> else
>     color.a = 1.0f - color.a;
> ```

Actually, this isn't right either. Let me have a play.

Reply:
So, the filtered blend isn't going to work unless we have the background color available in the shader. The filtering is multiplicative, which means just using 0,0,0,1 as the initial color forces the first color to multiply by zero, resulting in zero (obviously). If you're happy to accept a fixed color background, it could work, otherwise perhaps it's just a step too far for this sample. It would require a multi-pass approach to work against any background sadly.

Reply:
Ah too bad!
I removed the filter blending option, because it would probably stray too far from the intent of the sample, which is mostly to use the per-pixel linked lists for OIT.